File: backend/src/modules/cart/cart.controller.js
import asyncHandler from 'express-async-handler';
import ApiResponse from '../../core/utils/api_response.util.js';
import CartService from './cart.service.js';

const addToCart = asyncHandler(async (req, res) => {
  const userId = req.user.id;
  const { productId } = req.body;
  const quantity = req.body.quantity || 1;

  const result = await CartService.addToCart(userId, productId, quantity);

  return res.status(200).json(new ApiResponse(200, result.message, null));
});

const getCart = asyncHandler(async (req, res) => {
  const userId = req.user.id;

  const cartItems = await CartService.getCart(userId);

  return res.status(200).json(new ApiResponse(200, 'Cart retrieved successfully', cartItems));
});

const removeFromCart = asyncHandler(async (req, res) => {
  const userId = req.user.id;
  const { productId } = req.body;

  const result = await CartService.removeFromCart(userId, productId);

  return res.status(200).json(new ApiResponse(200, result.message, null));
});

const updateCartQuantity = asyncHandler(async (req, res) => {
  const userId = req.user.id;
  const { productId, quantity } = req.body;

  const result = await CartService.updateCartQuantity(userId, productId, quantity);

  return res.status(200).json(new ApiResponse(200, result.message, null));
});

const clearCart = asyncHandler(async (req, res) => {
  const userId = req.user.id;

  const result = await CartService.clearCart(userId);

  return res.status(200).json(new ApiResponse(200, result.message, null));
});

const getCartCount = asyncHandler(async (req, res) => {
  const userId = req.user.id;

  const count = await CartService.getCartCount(userId);

  return res.status(200).json(new ApiResponse(200, 'Cart count retrieved successfully', { count }));
});

export { addToCart, getCart, removeFromCart, updateCartQuantity, clearCart, getCartCount };

File: frontend/lib/modules/user/cart/controller/cart_controller.dart
import 'package:frontend/core/utils/api_client.dart';
import 'package:frontend/modules/user/cart/models/cart.dart';
import 'package:get/get.dart';

class CartController extends GetxController {
  Rx<bool> isLoading = false.obs;
  Rx<bool> isFetching = false.obs;
  Rx<String> errorMessage = ''.obs;
  Rx<String> responseMessage = ''.obs;
  RxList<Cart> cartItems = <Cart>[].obs;

  Future<bool> addToCart(String productId, int quantity) async {
    try {
      errorMessage.value = '';
      final response = await apiClient.post(
        'cart/add',
        data: {'productId': productId, 'quantity': quantity},
      );
      responseMessage.value = response['message'];
      return true;
    } catch (e) {
      errorMessage.value = e.toString();
      return false;
    }
  }

  Future<List<Cart>> getCart() async {
    try {
      isFetching.value = true;
      errorMessage.value = '';
      final response = await apiClient.get('cart/all');

      if (response['data'] is List) {
        final data = response['data'] as List;
        cartItems.value = data.map((e) => Cart.fromJson(e)).toList();
      } else {
        cartItems.value = [];
      }

      return cartItems;
    } catch (e) {
      errorMessage.value = e.toString();
      return [];
    } finally {
      isFetching.value = false;
    }
  }

  Future<bool> removeFromCart(String productId) async {
    try {
      errorMessage.value = '';
      final response = await apiClient.post('cart/remove', data: {'productId': productId});
      responseMessage.value = response['message'];

      // Remove from local list
      cartItems.removeWhere((item) => item.productId == productId);

      return true;
    } catch (e) {
      errorMessage.value = e.toString();
      return false;
    }
  }

  Future<bool> updateCartQuantity(String productId, int quantity) async {
    try {
      errorMessage.value = '';
      final response = await apiClient.put(
        'cart/update-quantity',
        data: {'productId': productId, 'quantity': quantity},
      );
      responseMessage.value = response['message'];

      // Update local list
      final index = cartItems.indexWhere((item) => item.productId == productId);
      if (index != -1) {
        cartItems[index].quantity = quantity;
        cartItems.refresh();
      }

      return true;
    } catch (e) {
      errorMessage.value = e.toString();
      return false;
    }
  }

  Future<bool> clearCart() async {
    try {
      errorMessage.value = '';
      final response = await apiClient.post('cart/clear', data: {});
      responseMessage.value = response['message'];

      // Clear local list
      cartItems.value = [];

      return true;
    } catch (e) {
      errorMessage.value = e.toString();
      return false;
    }
  }

  Future<int> getCartCount() async {
    try {
      errorMessage.value = '';
      final response = await apiClient.get('cart/count');
      final count = response['data']['count'] as int;
      return count;
    } catch (e) {
      errorMessage.value = e.toString();
      return 0;
    }
  }
}
